pkg/codegen/docs/gen.go

```diff
@@ -1753,167 +1757,142 @@ func (mod *modContext) getModuleFileName() string {
+	return moduleConflictResolver{
+		dctx: dctx,
+		seen: map[string]struct{}{},
+	}
 }
 
-func sortIndexEntries(entries []indexEntry) {
-	if len(entries) == 0 {
-		return
+// getSafeName returns a documentation name for an item
+// that is unique within the module.
+//
+// if the item has already been resolved by any module,
+// the previously-resolved name is returned.
+func (r *moduleConflictResolver) getSafeName(name string, item interface{}) string {
+	if safeName, ok := r.dctx.moduleConflictLinkMap[item]; ok {
+		return safeName
 	}
 
-	sorter := &indexEntrySorter{
-		entries: entries,
+	var prefixes []string
+	switch item.(type) {
+	case *schema.Resource:
+		prefixes = []string{"", "res-"}
+	case *schema.Function:
+		prefixes = []string{"", "fn-"}
+	case *modContext:
+		prefixes = []string{"", "mod-"}
+	default:
+		prefixes = []string{""}
+	}
+	for _, prefix := range prefixes {
+		candidate := prefix + name
+		if _, exists := r.seen[candidate]; exists {
+			continue
+		}
+		r.seen[candidate] = struct{}{}
+		r.dctx.moduleConflictLinkMap[item] = candidate
+		return candidate
 	}
 
-	sort.Sort(sorter)
+	glog.Error("skipping unresolvable duplicate file name: ", name)
+	return ""
 }
 
-// genIndex emits an _index.md file for the module.
-func (mod *modContext) genIndex() indexData {
+func (mod *modContext) gen(fs codegen.Fs) error {
 	glog.V(4).Infoln("genIndex for", mod.mod)
-	modules := make([]indexEntry, 0, len(mod.children))
-	resources := make([]indexEntry, 0, len(mod.resources))
-	functions := make([]indexEntry, 0, len(mod.functions))
 
 	modName := mod.getModuleFileName()
-	title := modName
+	conflictResolver := mod.docGenContext.newModuleConflictResolver()
 
 	def, err := mod.pkg.Definition()
 	contract.AssertNoErrorf(err, "failed to get definition for package %q", mod.pkg.Name())
 
-	// An empty string indicates that this is the root module.
-	if title == "" {
+	modTitle := modName
+	if modTitle == "" {
+		// An empty string indicates that this is the root module.
 		if def.DisplayName != "" {
-			title = def.DisplayName
+			modTitle = def.DisplayName
 		} else {
-			title = getPackageDisplayName(mod.pkg.Name())
+			modTitle = getPackageDisplayName(mod.pkg.Name())
 		}
 	}
 
+	// addFileTemplated executes template tmpl with data,
+	// and adds a file $dirName/_index.md with the result.
+	addFileTemplated := func(dirName, tmpl string, data interface{}) error {
+		var buff bytes.Buffer
+		if err := mod.docGenContext.templates.ExecuteTemplate(&buff, tmpl, data); err != nil {
+			return err
+		}
+		p := path.Join(modName, dirName, "_index.md")
+		fs.Add(p, buff.Bytes())
+		return nil
+	}
+
 	// If there are submodules, list them.
+	modules := make([]indexEntry, 0, len(mod.children))
 	for _, mod := range mod.children {
 		modName := mod.getModuleFileName()
 		displayName := modFilenameToDisplayName(modName)
+		safeName := conflictResolver.getSafeName(displayName, mod)
+		if safeName == "" {
+			continue // unresolved conflict
```

Comment:
hmm I know this was @abhinav comment but I was more thinking to err on the current behavior of if we get an unresolved conflict write out with the unsafe only differing by case name.
Just means if we do hit this we write out what we used to write out, rather than nothing.
Having said that the chance of hitting this feels _very low_ so I'm not overly fussed about changing it unless someone thinks my very low probability is actually wrong.